### DISCUSSION TITLE
Can only keeping the top level functions and deleting the function's internal AST reduce memory usage for large projects?

### DISCUSSION THREAD

**Discussion by aerosayan** [clangd/clangd#1647] [Ideas]
Hello everyone,

**In my opinion go-to-def is the most useful and used features of clangd.**

clangd has improved a lot, and thanks to the dev team for their efforts.

But still, the main issue many users seem to have, is high memory usage.

Opening even small 4-5 c++ file containing template code increases memory usage to 400 MB or more.

This seems very high.

I have disabled the background indexer, and that has helped little bit. I'm trying to disable linting to see if that reduces memory usage, but I currently don't know how to do that.

So, I wanted to ask if there's any way to reduce memory consumption.

Kindly correct me if I'm wrong, but it seems like clangd is storing the full AST for almost all of the functions in the opened/included files.

My idea is: **What if clangd only stored the AST of only the global symbols and the top-level function declarations/definitions,  while everything inside the functions' AST was deleted?**

For example, what if inside clangd's AST ...

```c++
int main()
{
    std::cout << "hello world" << std::endl;
    return 0;
}
```

was reduced down to the AST of ...

```c++
int main()
{
}
```

This is a simplified example, but this pruning could be applied to all heavy templated functions too.

**If this is possible, this would be extremely fast for clangd to compile, and require extremely small amount of memory to store. This will allow clangd to be useful as an LSP on very large projects like Linux Kernel.**

But I understand that this method has some limitations.

For example:
1. We can't rename symbols globally if the full AST is not loaded.
2. We can't have linting if the full AST is not loaded.
3. We can't go-to-def onto different variables inside a function.

In my humble opinion, it's an acceptable trade-off., because the most important and useful feature of clangd is go-to-def for jumping quickly across multiple files in a large project.

Is this possible?

Thanks for all your efforts!

**Comment by HighCommander4** [comment]
Clangd already skips parsing the bodies of functions in included files.

It parses the bodies of functions that appear directly in an open file (as opposed in a header included from an open file), but that seems necessary to provide editor functionality (for example, when you perform go-to-definition on a function, presumably in many cases you're jumping **from** a call that's inside a function body, so clangd needs to have an AST for that function body to recognize the call).

**Reply by aerosayan** [reply]
Thank you for your quick reply and sharing how clangd works.

I've found a few different ways to reduce clangd memory usage in neovim, and have a few more question/suggestion if it's not inconvenience to you. 

For example, neovim keeps many open buffers even if we quit them using `:q`, and that keeps the clangd server running in the background for all those files. So memory consumption skyrockets. The only way I have found to make clangd free the memory for closed files, is to close them with `:bd` (buffer-delete) instead of `:q`.

_**Maybe clangd team could inform neovim team about this behavior,** because if many user's laptops become unresponsive while using neovim+clangd, and runs out of memory (OOM crash), then the user will most likely think clangd is at fault. While in reality, neovim is more responsible for this OOM crash because of not freeing the resources correctly._

Additionally, obviously clangd has  a lot of complicated features to support, so little bit high memory usage is expected, but I'm just wondering if it would be possible to limit the memory used  to a certain amount (say 500 MB)?

**_Is it possible for the dev team to keep only few files open in memory, or small portions of the AST in memory?_**

For example: 

- if there are 5 files open, and the memory usage exceeds 500 MB (user specified limit), then clangd will close few of the files until the memory used drops below the limit. 
- if there are 10 headers being included, maybe it's not necessary to load all of them in memory.
- if a file is very large, maybe clangd could only load the AST of a single function the user is currently in. The AST of the opened file could be written to disk, and only the current function's AST could be streamed in when necessary.

Disk IO would most likely be little bit slow and make the editor+clangd little jittery, but that's acceptable if we can use clangd on large projects on a standard laptop (4GB or 8GB RAM).

Based on the resources available, the user could then specify the limit through the `.clangd` configuration file, and clangd would become extremely responsive even on large projects. 

Thanks.

**Reply by tom-anders** [reply]
> For example, neovim keeps many open buffers even if we quit them using `:q`, and that keeps the clangd server running in the background for all those files. So memory consumption skyrockets. The only way I have found to make clangd free the memory for closed files, is to close them with `:bd` (buffer-delete) instead of `:q`.
> 

There are Plugins out there (don't know a name of the top of my head) which automatically close inactive buffers after a certain amount of time, maybe you might want to look into that

**Reply by HighCommander4** [reply]
Clangd's behaviour is terms of which files it considers to be open is pretty clear in terms of the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/) specification: clangd opens a file (creating an AST for it) when the client sends a `textDocument/didOpen` notification, and closes it when the client sends a `textDocument/didClose` notification.

The question of how a client maps operations like `:q` or `:bd` onto LSP notifications is a matter for the client -- if you'd like neovim to send `textDocument/didClose` on `:q`, I would suggest filing that issue with your neovim LSP client.

(Also, while I'm not a neovim user, I understand that there are multiple LSP clients available for it, such as [coc](https://github.com/neoclide/coc.nvim), [YouCompleteMe](https://ycm-core.github.io/YouCompleteMe/), and [LanguageClient-neovim](https://github.com/autozimu/LanguageClient-neovim). Perhaps one of these already has the behaviour your desire.)

**Reply by HighCommander4** [reply]
On the more general topic of limiting clangd's memory usage, here's some background information of what might be taking up clangd's memory:

 * **Preambles**. This is the part of the AST of each open file that represents the contents of #included headers. It tends to be the largest part of the AST (compared to the main-file part).
   * **This is already stored on disk**, unless you're running clangd with `--pch-storage=memory` (which is not the default). With the default value of `--pch-storage=disk`, I would expect the memory usage of preambles to be fairly low (I believe the only thing kept in memory would be specific declarations that are deserialized from disk as needed to satisfy editor queries).
 * **Main-file ASTs**. This is the part of the AST that represents declarations contained directly in open files.
   * Clangd doesn't actually retain these for all open files, it only keeps the most recently used [**3**](https://searchfox.org/llvm/rev/49946df8211e9d36f0b3755e64b55bc28c0a4247/clang-tools-extra/clangd/TUScheduler.h#65) and discards older ones. (We can do this because main-file ASTs are quick to rebuild compared to preambles.)
   * I suppose we could make this cache size of 3 configurable, though I think 3 is already pretty low.
 * **The project index**. This is a database of information about the project that clangd builds in the background by parsing all of the files in the project, used to answer requests that require cross translation unit information, like "find all references".
   * This is stored on disk, but clangd reads it all into memory on startup. For large projects, this is often the biggest contributor to clangd's memory usage, as its size is not related to the number of **open** files, but the size of the entire project.
   * The index is optional, it can be disabled by running clangd with `--background-index=false` (but then cross translation unit features won't work).
   * There is also a possibility of consuming an index served by a remote machine, see https://clangd.llvm.org/guides/remote-index.
   * In principle, the index could be rearchitected in a way that allows most of it contents to be paged in from disk as needed, instead of being loaded all into memory, but this would involve a nontrivial engineering effort. Some discussion about this can be found in https://github.com/clangd/clangd/discussions/1632.

Clangd supports a protocol extension to report the breakdown of its current [memory usage](https://clangd.llvm.org/extensions#memory-usage). You may want to use a client that supports this (some that I'm aware of are [vscode-clangd](https://github.com/clangd/vscode-clangd), [coc-clangd](https://github.com/clangd/coc-clangd), and [qt-creator](https://github.com/qt-creator/qt-creator)) to check where most of the memory usage is coming from in your case, so you know what to focus on.
